### ISSUE TITLE
Duplicate attribution callback invocation

### ISSUE THREAD

**Issue by sergiymomot** [adjust/ios_sdk#518]
3 days ago, we noticed that attribution callback is called when there was no fresh install or reinstall of our app.
It happened at around March 23rd 8pm UTC, we saw a spike of attribution callbacks for existing users who were already attributed before.
For my personal device (there was no app reinstall), I received 5 attribution callbacks during one day with the same Adjust tracker (Organic). Based on tracking, majority of our iOS users got re-attributed again.

On attribution callback being invoked, we track _Install Attributed_ event like this:
```
- (void)adjustAttributionChanged:(ADJAttribution *)attribution
{
    [self.analytics track:@"Install Attributed" properties:@{
        @"provider" : @"Adjust",
        @"trackerToken" : attribution.trackerToken ?: [NSNull null],
        @"trackerName" : attribution.trackerName ?: [NSNull null],
        @"campaign" : @{
            @"source" : attribution.network ?: [NSNull null],
            @"name" : attribution.campaign ?: [NSNull null],
            @"content" : attribution.clickLabel ?: [NSNull null],
            @"adCreative" : attribution.creative ?: [NSNull null],
            @"adGroup" : attribution.adgroup ?: [NSNull null]
        }
    }];
}
```

I looked at SDK code, and noticed that SDK performs de-duplication by storing previous attribution on disk, and later, when new attribution is received, it compares new data to previously stored one and if it is the same, then delegate callback is not being called. 
And this worked perfectly for us for over 2 years, until recently de-duplication stopped working as we started to see 10+ attribution callbacks being called for some users with the same parameters tracked above.

Since `isEqual` method that compares ADJAttribution objects is unlikely to suddenly break, the only thing that comes to my mind is that Adjust Device ID is being regenerated for people for no apparent reason, and then during comparison of attribution objects, `isEqual` returns false because Adjust Device ID has changed. That would imply the server change, and would explain why it started simultaneously on all our app versions some of which are more than 3 months old.

Hope you can clarify for me why attribution callback might be called multiple times when no attribution changed, and if it is possible that Adjust Device ID calculation was recently changed that caused re-attribution of most our existing users.

Thanks

**Comment by sergiymomot** [comment]
Found the issue of duplicate attribution callback invocation:

```
- (BOOL)isEqualToAttribution:(ADJAttribution *)attribution {
    if (attribution == nil) {
        return NO;
    }
    if (![NSString adjIsEqual:self.trackerToken toString:attribution.trackerToken]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.trackerName toString:attribution.trackerName]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.network toString:attribution.network]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.campaign toString:attribution.campaign]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.adgroup toString:attribution.adgroup]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.creative toString:attribution.creative]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.clickLabel toString:attribution.clickLabel]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.adid toString:attribution.adid]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.costType toString:attribution.costType]) {
        return NO;
    }
    // costAmount is of type NSNumber
    if (![self.costAmount isEqualToNumber:attribution.costAmount]) {
        return NO;
    }
    if (![NSString adjIsEqual:self.costCurrency toString:attribution.costCurrency]) {
        return NO;
    }

    return YES;
}

```

This method is called whenever Adjust SDK receives new attribution data from server and compares it to previous locally stored data. The problem is in costAmount comparison: `isEqualToNumber` method returns false if both NSNumber objects are `nil`.

This was always an issue I believe, however it seems that up until 8pm UTC March 23rd, servers were returning 0 for costAmount rather than null, so comparing 0s worked properly.
Again, the issue started in all our iOS app versions at the same time which have multiple versions of Adjust SDKs.

Can you confirm my guess of servers changing costAmount from 0 to null if no cost is present for install attribution?

**Comment by Awakening7520** [comment]
?????
Qué me pasó? Todavía estoy aquí abajo con la gente y tú estás en la casa de tu tío que te está haciendo el otro por ahí abajo con tu coche y yo no te puedo ayudar con tu tío ✅ te quiero mucho y tú dime algo Te quiero un montón

**Comment by uerceg** [comment]
@sergiymomot Thanks for reporting this, fix is on its way.

**Comment by uerceg** [comment]
We just released [v4.27.1](https://github.com/adjust/ios_sdk/releases/tag/v4.27.1) which should fix this behaviour. In case you still face this issue or have any further question, feel free to reopen/comment.

Thank you one more time for reporting this!

Cheers

**Comment by T-10810** [comment]
So good so far
